samples/synchronization/src/main.c

```diff
@@ -16,6 +16,8 @@
  * world application would likely use the static approach for both threads.
  */
 
+#define PIN_THREADS (defined(CONFIG_SMP) && defined(CONFIG_SCHED_CPU_MASK) \
```

Comment:
Apparently newer LLVMs complain if a macro that expands to `defined()` instances is used as a preprocessor expression.  See #32078.  Better to do:
```
#if defined(CONFIG_SMP) && defined(CONFIG_SCHED_CPU_MASK) && CONFIG_MP_NUM_CPUS > 1)
#define PIN_THREADS 1
#else
#define PIN_THREADS 0
#endif
```

Reply:
so this is enabled by default in llvm but not in gcc, we need to sync both and add either 

`-Wexpansion-to-defined` to gcc or `-Wno-expansion-to-defined` to llvm to avoid this from going in.

The question if we want to allow this in general or if we should disallow such usage of defined().

```
'-Wexpansion-to-defined'
     Warn whenever 'defined' is encountered in the expansion of a macro
     (including the case where the macro is expanded by an '#if'
     directive).  Such usage is not portable.  This warning is also
     enabled by '-Wpedantic' and '-Wextra'.
```

"not portable" seems a like a reason good enough to disallow it.

@tejlmand FYI

Reply:
I'm fine with disabling it.  AFAICT the littlefs one was the only in-tree use of this "feature".

Reply:
@nashif see you fixed it here: https://github.com/zephyrproject-rtos/zephyr/pull/32330